File: android-im-ui/src/main/kotlin/com/flare/im/ui/GroupCallView.kt
package com.flare.im.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.MicOff
import androidx.compose.material.icons.filled.VideocamOff
import androidx.compose.material.icons.outlined.ExpandMore
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/** Group (multi-party) call — participant grid + controls. Spec: Call/GroupCallView. */
@Composable
fun GroupCallView(
    participants: List<CallParticipant>,
    mode: FlareCallMode,
    state: String,
    title: String? = null,
    durationLabel: String? = null,
    muted: Boolean = false,
    cameraOn: Boolean = true,
    speakerOn: Boolean = false,
    onHangup: (() -> Unit)? = null,
    onToggleMute: (() -> Unit)? = null,
    onToggleCamera: (() -> Unit)? = null,
    onToggleSpeaker: (() -> Unit)? = null,
    onSwitchCamera: (() -> Unit)? = null,
    onMinimize: (() -> Unit)? = null,
) {
    val cols = when {
        participants.size <= 1 -> 1
        participants.size <= 4 -> 2
        participants.size <= 9 -> 3
        else -> 4
    }
    val strings = flareStrings()
    val status = when (state) {
        "connected" -> durationLabel ?: strings.callConnected
        "ringing" -> strings.callRinging
        else -> strings.callCalling
    }
    Column(
        Modifier.fillMaxWidth().background(
            Brush.verticalGradient(listOf(Color(0xFF211D30), Color(0xFF17131F), Color(0xFF100C17)))),
    ) {
        Row(Modifier.fillMaxWidth().padding(start = 16.dp, end = 16.dp, top = 14.dp, bottom = 4.dp),
            verticalAlignment = Alignment.CenterVertically) {
            Box(Modifier.size(36.dp).clip(CircleShape).background(Color.White.copy(alpha = 0.12f))
                .then(if (onMinimize != null) Modifier.clickable { onMinimize() } else Modifier),
                contentAlignment = Alignment.Center) {
                Icon(Icons.Outlined.ExpandMore, contentDescription = null, tint = Color.White, modifier = Modifier.size(20.dp))
            }
            Spacer(Modifier.width(12.dp))
            Column {
                Text(title ?: flareStrings().groupCall, color = Color.White, fontWeight = FontWeight.SemiBold,
                    fontSize = 16.sp, maxLines = 1, overflow = TextOverflow.Ellipsis)
                Text(flareStrings().joinedCount(participants.size, status), color = Color.White.copy(alpha = 0.62f), fontSize = 12.sp)
            }
        }
        LazyVerticalGrid(
            columns = GridCells.Fixed(cols),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
            contentPadding = PaddingValues(horizontal = 12.dp, vertical = 4.dp),
            modifier = Modifier.weight(1f),
        ) {
            items(participants, key = { it.id }) { p -> callTile(p, mode) }
        }
        Box(Modifier.fillMaxWidth().padding(vertical = 20.dp), contentAlignment = Alignment.Center) {
            CallControls(muted = muted, cameraOn = cameraOn, speakerOn = speakerOn, mode = mode,
                onToggleMute = onToggleMute, onToggleCamera = onToggleCamera, onToggleSpeaker = onToggleSpeaker,
                onSwitchCamera = onSwitchCamera, onHangup = onHangup)
        }
    }
}

@Composable
private fun callTile(p: CallParticipant, mode: FlareCallMode) {
    Box(
        Modifier.aspectRatio(0.86f).clip(RoundedCornerShape(16.dp))
            .background(if (p.isSelf) Color(0x297C3AED) else Color.White.copy(alpha = 0.06f))
            .border(2.dp, if (p.speaking) Color(0xFF34D17F) else Color.Transparent, RoundedCornerShape(16.dp)),
        contentAlignment = Alignment.Center,
    ) {
        Avatar(userId = p.id, displayName = p.name, size = 56.dp)
        Row(Modifier.align(Alignment.BottomStart).padding(8.dp), verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(5.dp)) {
            if (p.muted) {
                Box(Modifier.size(20.dp).clip(RoundedCornerShape(6.dp)).background(Color.Black.copy(alpha = 0.42f)),
                    contentAlignment = Alignment.Center) {
                    Icon(Icons.Filled.MicOff, contentDescription = null, tint = Color.White, modifier = Modifier.size(12.dp))
                }
            } else if (p.cameraOff && mode == FlareCallMode.Video) {
                Box(Modifier.size(20.dp).clip(RoundedCornerShape(6.dp)).background(Color.Black.copy(alpha = 0.42f)),
                    contentAlignment = Alignment.Center) {
                    Icon(Icons.Filled.VideocamOff, contentDescription = null, tint = Color.White, modifier = Modifier.size(12.dp))
                }
            }
            Text(if (p.isSelf) flareStrings().selfSuffix(p.name) else p.name, color = Color.White, fontSize = 12.sp,
                maxLines = 1, overflow = TextOverflow.Ellipsis,
                modifier = Modifier.clip(RoundedCornerShape(6.dp)).background(Color.Black.copy(alpha = 0.42f))
                    .padding(horizontal = 8.dp, vertical = 2.dp))
        }
    }
}

File: ios-im-ui/Sources/FlareIMUI/Components/GroupCallView.swift
import SwiftUI

// MARK: - GroupCallView

/// Group (multi-party) call — participant grid + controls.
/// Spec: Call/GroupCallView (`GroupCallView`).
public struct GroupCallView: View {
    private let participants: [CallParticipant]
    private let mode: FlareCallMode
    private let state: FlareCallState
    private let title: String?
    private let durationLabel: String?
    private let muted: Bool
    private let cameraOn: Bool
    private let speakerOn: Bool
    private let onHangup: (() -> Void)?
    private let onToggleMute: (() -> Void)?
    private let onToggleCamera: (() -> Void)?
    private let onToggleSpeaker: (() -> Void)?
    private let onSwitchCamera: (() -> Void)?
    private let onMinimize: (() -> Void)?

    public init(participants: [CallParticipant], mode: FlareCallMode, state: FlareCallState,
                title: String? = nil, durationLabel: String? = nil, muted: Bool = false, cameraOn: Bool = true,
                speakerOn: Bool = false, onHangup: (() -> Void)? = nil, onToggleMute: (() -> Void)? = nil,
                onToggleCamera: (() -> Void)? = nil, onToggleSpeaker: (() -> Void)? = nil,
                onSwitchCamera: (() -> Void)? = nil, onMinimize: (() -> Void)? = nil) {
        self.participants = participants; self.mode = mode; self.state = state; self.title = title
        self.durationLabel = durationLabel; self.muted = muted; self.cameraOn = cameraOn; self.speakerOn = speakerOn
        self.onHangup = onHangup; self.onToggleMute = onToggleMute; self.onToggleCamera = onToggleCamera
        self.onToggleSpeaker = onToggleSpeaker; self.onSwitchCamera = onSwitchCamera; self.onMinimize = onMinimize
    }

    private var cols: Int {
        let n = participants.count
        if n <= 1 { return 1 }; if n <= 4 { return 2 }; if n <= 9 { return 3 }; return 4
    }
    private var statusText: String {
        if state == .connected { return durationLabel ?? "已接通" }
        if state == .ringing { return "响铃中…" }
        return "呼叫中…"
    }

    public var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 12) {
                Button { onMinimize?() } label: {
                    Image(systemName: "chevron.down").font(.system(size: 20)).foregroundColor(.white)
                        .frame(width: 36, height: 36).background(Circle().fill(Color.white.opacity(0.12)))
                }.buttonStyle(.plain)
                VStack(alignment: .leading, spacing: 2) {
                    Text(title ?? "群通话").font(.system(size: 16, weight: .semibold)).foregroundColor(.white).lineLimit(1)
                    Text("\(participants.count) 人已加入 · \(statusText)").font(.system(size: 12)).foregroundColor(.white.opacity(0.62))
                }
                Spacer()
            }.padding(.horizontal, 16).padding(.top, 14).padding(.bottom, 4)

            let grid = Array(repeating: GridItem(.flexible(), spacing: 8), count: cols)
            ScrollView {
                LazyVGrid(columns: grid, spacing: 8) {
                    ForEach(participants) { p in tile(p) }
                }.padding(.horizontal, 12).padding(.top, 4)
            }

            CallControlsView(muted: muted, cameraOn: cameraOn, speakerOn: speakerOn, mode: mode,
                             onToggleMute: onToggleMute, onToggleCamera: onToggleCamera,
                             onToggleSpeaker: onToggleSpeaker, onSwitchCamera: onSwitchCamera, onHangup: onHangup)
                .padding(.vertical, 20).padding(.bottom, 16)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(
            LinearGradient(colors: [Color(.sRGB, red: 0.129, green: 0.114, blue: 0.188, opacity: 1),
                                    Color(.sRGB, red: 0.09, green: 0.075, blue: 0.122, opacity: 1),
                                    Color(.sRGB, red: 0.063, green: 0.047, blue: 0.09, opacity: 1)],
                           startPoint: .top, endPoint: .bottom)
        )
    }

    private func tile(_ p: CallParticipant) -> some View {
        ZStack(alignment: .bottomLeading) {
            RoundedRectangle(cornerRadius: 16)
                .fill(p.isSelf ? Color(.sRGB, red: 0.486, green: 0.227, blue: 0.929, opacity: 0.16) : Color.white.opacity(0.06))
                .overlay(RoundedRectangle(cornerRadius: 16).stroke(p.speaking ? Color(.sRGB, red: 0.204, green: 0.82, blue: 0.498, opacity: 1) : .clear, lineWidth: 2))
                .aspectRatio(0.86, contentMode: .fit)
            AvatarView(userId: p.id, displayName: p.name, avatarURL: p.avatarURL, size: 56)
                .frame(maxWidth: .infinity, maxHeight: .infinity)
            HStack(spacing: 5) {
                if p.muted {
                    Image(systemName: "mic.slash.fill").font(.system(size: 10)).foregroundColor(.white)
                        .frame(width: 20, height: 20).background(RoundedRectangle(cornerRadius: 6).fill(Color.black.opacity(0.42)))
                } else if p.cameraOff && mode == .video {
                    Image(systemName: "video.slash.fill").font(.system(size: 10)).foregroundColor(.white)
                        .frame(width: 20, height: 20).background(RoundedRectangle(cornerRadius: 6).fill(Color.black.opacity(0.42)))
                }
                Text(p.isSelf ? "\(p.name)（我）" : p.name).font(.system(size: 12)).foregroundColor(.white).lineLimit(1)
                    .padding(.horizontal, 8).padding(.vertical, 2)
                    .background(RoundedRectangle(cornerRadius: 6).fill(Color.black.opacity(0.42)))
            }.padding(8)
        }
        .clipShape(RoundedRectangle(cornerRadius: 16))
    }
}
